### PR TITLE
Feature/react adaptive animation

### PR DESCRIPTION
**Pull request by anton-karlovskiy**
For now, it's not ready yet to be merged.

TODO:
1. One ask for the animation project would be including just a summary in the README so we can point folks to where the animations are being toggled on/off in source. Perhaps good for when you are PR'ing back to the adaptive-loading repo. -- done
2. How to deploy Next.js project to adaptive loading firebase instance following the previous projects. -- done by workaround
3. The main difference would be referencing the memory hook that is in the adaptive-loading repo when you are working on the PR vs us having a separate implementation. -- issue
4. What if we detected when Next is executing on the server-side and use Client Hints for the signals there but if you are client side we rely on the JavaScript signals? -- done
5. HOC and hardware concurrency might be considered.
6. License attaching. -- done

### CODE REVIEW

**Comment by addyosmani** [comment]
Ruh roh. After a fresh checkout and install of the dependencies, I'm running into some challenges with `npm run start`/`yarn start`. Without changing any settings, tapping any of the stories results in a route navigation failure:



image

**Comment by anton-karlovskiy** [comment]
@addyosmani 
Whoops, sorry.
There was a mistake in comparison between string and number.
I fixed and pushed again. Please try it again.

**Comment by addyosmani** [comment]
Thanks! I'm now able to navigate to routes when animation is on. When animation is off, I'm still seeing this bug:

image
image

**Comment by anton-karlovskiy** [comment]
@addyosmani 
Let me check right away.

**Comment by anton-karlovskiy** [comment]
@addyosmani 
Sorry, please try again now. It will be working. There was another comparison I missed.

File: cna-memory-considerate-animation/pages/_app.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import { CLIENT_HINT_MEMORY_LIMIT } from '../config';
+
+const MyApp = ({ Component, pageProps, router }) => {
+  const [manualEnabled, setManualEnabled] = useState(false);
+  const [isAnimationOn, setIsAnimationOn] = useState(true);
+
+  const { clientHintDeviceMemory, memoryStatus } = pageProps;
+  let overLoaded;
+  if (clientHintDeviceMemory) {
+    overLoaded = clientHintDeviceMemory < CLIENT_HINT_MEMORY_LIMIT;
```
Comment (anton-karlovskiy) [inline cna-memory-considerate-animation/pages/_app.js]: Memory status evaluation logic with Device Memory Client Hint on server side rendering.

File: cna-memory-considerate-animation/utils/hooks.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    const usedMemoryPercent = getUsedMemoryPercent();
+    let overLoaded = false;
+    // Check if we've exceeded absolute memory limit
+    if (overUsedMemorySize > 0) {
+      overLoaded = true;
+    }
+    // Check if we've exceeded relative memory limit for client
+    if (usedMemoryPercent > MAX_PERCENT_THRESHOLD) {
+      overLoaded = true;
+    }
```
Comment (anton-karlovskiy) [inline cna-memory-considerate-animation/utils/hooks.js]: Memory status evaluation logic with window.performance.memory on client side rendering.

File: cna-memory-considerate-animation/pages/_app.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    console.log('[_app MyApp] Client Hint Device Memory based Memory Overloaded => ', overLoaded);
+  } else {
+    overLoaded = memoryStatus.overLoaded;
+    console.log('[_app MyApp] Memory React Hook based Memory Overloaded => ', overLoaded);
+  }
```
Comment (anton-karlovskiy) [inline cna-memory-considerate-animation/pages/_app.js]: Memory status evaluation conditional for server side rendering and client side rendering.

**Comment by addyosmani** [comment]
I tested the latest implementation and functionality appears to be as expected for the most part. Could you share what you see as being remaining here?

**Comment by anton-karlovskiy** [comment]
> I tested the latest implementation and functionality appears to be as expected for the most part. Could you share what you see as being remaining here?

As for the implementation and functionality, I think it's done.
The remainders are https://github.com/GoogleChromeLabs/adaptive-loading/pull/52#issue-323439433.
I'll get this done shortly.

**Comment by anton-karlovskiy** [comment]
@addyosmani
I prepared the README draft for this demo (subproject README) but I'm afraid I'm not sure where we should put this demo in the following category of the whole project README.
Please choose one for me. Thank you.

### Adaptive Loading patterns
### Data Saver-style patterns
### Full applications

**Comment by addyosmani** [comment]
Let's put this in adaptive loading patterns

**Comment by anton-karlovskiy** [comment]
@addyosmani 
`3. The main difference would be referencing the memory hook that is in the adaptive-loading repo when you are working on the PR vs us having a separate implementation.`

I tried to link the memory hook from the main repo to the react animation project.
But due to [this issue](https://github.com/facebook/react/issues/13991), it looks a bit difficult to do so because we use React in the project as well as in the hooks.
https://github.com/facebook/react/issues/13991#issuecomment-433630460
https://github.com/facebook/react/issues/13991#issuecomment-435135293

**Comment by anton-karlovskiy** [comment]
@addyosmani 
This project is ready to be merged once you review the codebase and my reports.

**Comment by addyosmani** [comment]
As this is just a demo, let's move ahead. I am very surprised to hear about that React limitation but we can circle back on a better workaround later.

**Pull request merged**